src/dune_pkg/lock_dir.ml

```diff
@@ -705,3 +705,32 @@ module Load_immediate = Make_load (struct
   end)
 
 let read_disk = Load_immediate.load
+
+let transitive_dependency_closure t start =
+  let missing_packages =
+    let all_packages_in_lock_dir = Package_name.Set.of_keys t.packages in
+    Package_name.Set.diff start all_packages_in_lock_dir
+  in
+  match Package_name.Set.is_empty missing_packages with
+  | false -> Error (`Missing_packages missing_packages)
+  | true ->
+    let to_visit = Queue.create () in
```

Comment:
There's a `Top_closure` module you could take a look at. Perhaps it's usable here.

Reply:
It does look handy. I had a look for something like this a while back but didn't find it (I searched for names like "dag"). I'll merge this as is but look at switching to `Top_closure` in a later change. One possible issue is that `Top_closure` doesn't allow for dependency cycles but my understanding is that opam allows dependencies cycles as long as some of the deps are marked as `post`. Is that right? I've been intentionally allowing for dependency cycles in the dependency graph traversals I've implemented thus far for this reason.

Reply:
We actually ignore post deps for precisely this reason. Dune's engine cannot allow for cycles so we can assume that they aren't an issue.